pkg/reconciler/ingress/lister.go

```diff
@@ -0,0 +1,176 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ingress
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+
+	"go.uber.org/zap"
+	"istio.io/client-go/pkg/apis/networking/v1alpha3"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/sets"
+	corev1listers "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"knative.dev/serving/pkg/apis/networking/v1alpha1"
+	istiolisters "knative.dev/serving/pkg/client/istio/listers/networking/v1alpha3"
+	"knative.dev/serving/pkg/network"
+	"knative.dev/serving/pkg/network/ingress"
+	"knative.dev/serving/pkg/network/status"
+)
+
+// NewStatusProber creates a new instance of status.Prober
+func NewStatusProber(
+	logger *zap.SugaredLogger,
+	gatewayLister istiolisters.GatewayLister,
+	endpointsLister corev1listers.EndpointsLister,
+	serviceLister corev1listers.ServiceLister,
+	readyCallback func(*v1alpha1.Ingress)) *status.Prober {
+	return status.NewProber(
+		logger,
+		&gatewayPodTargetLister{
+			logger:          logger,
+			gatewayLister:   gatewayLister,
+			endpointsLister: endpointsLister,
+			serviceLister:   serviceLister,
+		},
+		readyCallback,
+	)
+}
+
+type gatewayPodTargetLister struct {
+	logger *zap.SugaredLogger
+
+	gatewayLister   istiolisters.GatewayLister
+	endpointsLister corev1listers.EndpointsLister
+	serviceLister   corev1listers.ServiceLister
+}
+
+// probeTargets represents the target to probe.
+type probeTarget struct {
+	// urlTmpl is an url template to probe.
+	urlTmpl string
+	// targetPort is a port of Gateway pod.
+	targetPort string
+}
+
+func (l *gatewayPodTargetLister) ListProbeTargets(ctx context.Context, ing *v1alpha1.Ingress) (map[string]map[string]sets.String, error) {
+	results := map[string]map[string]sets.String{}
+
+	qualifiedGatewayNames := qualifiedGatewayNamesFromContext(ctx)
+
+	gatewayHosts := ingress.HostsPerVisibility(ing, qualifiedGatewayNames)
+	for gatewayName, hosts := range gatewayHosts {
+		gateway, err := l.getGateway(gatewayName)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get Gateway %q: %w", gatewayName, err)
+		}
+		targetsPerPod, err := l.listGatewayTargetsPerPods(gateway)
+		if err != nil {
+			return nil, fmt.Errorf("failed to list the probing URLs of Gateway %q: %w", gatewayName, err)
+		}
+		if len(targetsPerPod) == 0 {
+			continue
+		}
+		for ip, targets := range targetsPerPod {
+			if _, existed := results[ip]; !existed {
+				results[ip] = map[string]sets.String{}
+			}
+			for _, host := range hosts.List() {
```

Comment:
`List` is a sorted list. Do you need that here? If not you can just iterate the `hosts`

Reply:
Added comment.

Reply:
Added comment